### ISSUE TITLE
probe() behaviour is confusing

### ISSUE THREAD

**Issue by whot** [libratbag/libratbag#370]
Assigning everyone to start the discussion. Hello all! :)

I had a look at the `probe()` function that libratbag uses internally to get the drivers. It's a bit confusing and less-than ideal. A driver is supposed to call `ratbag_device_init_profiles()` which calls back into `.read_profile()`, then into `.read_button` and `.read_led`. This seems a bit confusing, i wonder if a 'flat' `probe()` would be more sensible.

In other words, `ratbag_device_init_profiles()` would simply set up the device with sane defaults (based on num_buttons/num_leds/...) and return immediately. Then the driver can continue to loop through the now-setup device and set the various special bits. This would make it a lot easier to understand and debug, at the cost of not looking as nifty because we're not using as many function pointers :)

Thoughts?

**Comment by phomes** [comment]
I am in favor of a flat probe. These days reading profiles/buttons/leds are only done during probe anyway. It is hard to guess that ratbag_device_init_profiles() will call back into the driver. Searching for the callers of a function will also give better results if they are not hidden behind the function pointers.

**Comment by amfern** [comment]
it is abit complicating reading through theses sections, it would be easier with flatter probe.

**Comment by bentiss** [comment]
Looks like nobody likes the kernel `.probe()` model then :)

Works for me too in that case.
